core/src/autogluon/core/models/ensemble/stacker_ensemble_model.py

```diff
@@ -88,7 +94,49 @@ def _initialize(self, **kwargs):
             stack_column_prefix: self.base_model_names[i] for i, stack_column_prefix in enumerate(self.stack_column_prefix_lst)
         }
 
-        self._add_stack_to_feature_metadata()
+        feature_metadata = self._remove_unused_stack_in_feature_metadata(feature_metadata=feature_metadata)
+        return feature_metadata
+
+    def _get_dynamic_max_base_models_per_type(self, X: pd.DataFrame):
+        num_rows = len(X)
+        if num_rows < 1000:
+            max_models_per_type = 1
+        elif num_rows < 5000:
+            max_models_per_type = 2
+        elif num_rows < 10000:
+            max_models_per_type = 3
+        elif num_rows < 15000:
+            max_models_per_type = 4
+        elif num_rows < 20000:
+            max_models_per_type = 5
+        elif num_rows < 25000:
+            max_models_per_type = 6
+        elif num_rows < 30000:
+            max_models_per_type = 7
+        elif num_rows < 35000:
+            max_models_per_type = 8
+        elif num_rows < 40000:
+            max_models_per_type = 9
+        elif num_rows < 45000:
+            max_models_per_type = 10
+        elif num_rows < 50000:
+            max_models_per_type = 11
+        else:
+            max_models_per_type = 12
```

Comment:
How did we arrive at 12?
Also, what is the reason behind changing `max_models_per_type` based on the no. of rows in the data?

Reply:
12 is an arbitrary stopping point, but the intuition is that 12 should be close to approximating uncapped, and will behave much much better than if it is left uncapped if a user uses HPO and fits 1000s of models as an example.

This logic is not optimal, but it is better than a fixed value. We can iterate on the exact thresholds and values in future PRs.

The reason for changing the value depending on num_rows is that doing this shows improvement over a fixed value on TabRepo simulations. The intuition is that larger `max_models_per_type` leads to more overfitting, and you can counter overfitting by having more data.